### ISSUE TITLE
Change the backbone of the SimSiam to ViT

### ISSUE THREAD

**Issue by s9021025292140** [lightly-ai/lightly#1570]
Could you please advise on how to change the backbone of the SimSiam example to ViT? (https://docs.lightly.ai/self-supervised-learning/tutorials/package/tutorial_simsiam_esa.html)

Additionally, for the DINO (https://docs.lightly.ai/self-supervised-learning/examples/dino.html) and AIM (https://docs.lightly.ai/self-supervised-learning/examples/aim.html) examples, which use the PascalVOC dataset, how can I modify them to use my own classification dataset (with multiple folders, each containing images for a specific class)?

Thanks!

**Comment by guarin** [comment]
Hi!

> Could you please advise on how to change the backbone of the SimSiam example to ViT? (https://docs.lightly.ai/self-supervised-learning/tutorials/package/tutorial_simsiam_esa.html)

If you use the newest [timm](https://github.com/huggingface/pytorch-image-models) version you have to do the following changes:
```
import timm
backbone = timm.create_model("vit_tiny_patch16_224")

num_ftrs = backbone.num_features

class SimSiam(Module):
    def forward(self, x):
        # get representations 
        f = self.backbone.forward_features(x)
        f = self.backbone.pool(f).flatten(start_dim=1)
        # code below should be the same
        ...

# and for the embedding part:
with torch.no_grad():
    for i, (x, _, fnames) in enumerate(dataloader_test):
        # move the images to the gpu
        x = x.to(device)
        # embed the images with the pre-trained backbone
        y = model.backbone.forward_features(x)
        y = model.backbone.pool(y).flatten(start_dim=1)
        # store the embeddings and filenames in lists
        embeddings.append(y)
        filenames = filenames + list(fnames)
```

I would also recommend to switch the optimizer from SGD to `AdamW`. You might have to adjust the hyperparameters also a bit to make the model train well. ViTs are usually a bit tricky to get working.

> Additionally, for the DINO (https://docs.lightly.ai/self-supervised-learning/examples/dino.html) and AIM (https://docs.lightly.ai/self-supervised-learning/examples/aim.html) examples, which use the PascalVOC dataset, how can I modify them to use my own classification dataset (with multiple folders, each containing images for a specific class)?

You can use:
```
from lightly.data import LightlyDataset

dataset = LightlyDataset("path/to/folder")
```

**Comment by s9021025292140** [comment]
Thank you for your prompt response to my questions! I used DINO and loaded the pre-trained ViT small8 as the backbone for self-supervised learning in my training. Then, I used the following code for the downstream classification task:
```
import torch
import torch.nn as nn
class Classifier(torch.nn.Module):
    def __init__(self, backbone):
        super().__init__()
        # use the pretrained ResNet backbone
        self.backbone = backbone

        # freeze the backbone
        deactivate_requires_grad(backbone)

        # create a linear layer for our downstream classification model
        self.fc = nn.Linear(512, 10)

    def forward(self, x):
        y_hat = self.backbone(x).flatten(start_dim=1)
        y_hat = self.fc(y_hat)
        return y_hat

classifier = Classifier(backbone)
criterion = torch.nn.CrossEntropyLoss()
optimizer = torch.optim.SGD(classifier.fc.parameters(), lr=30.)
scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, max_epochs)

for epoch in range(100):
    print('epoch', epoch)

    # train
    classifier.train()
    train_loss = 0
    for x, y, _ in dataloader_train_classifier:
        optimizer.zero_grad()
        y_hat = classifier(x)
        loss = criterion(y_hat, y)
        loss.backward()
        optimizer.step()
        train_loss += loss.item() * len(x)
    train_loss = train_loss / len(dataloader_train_classifier.dataset)
    print('train loss', train_loss)

    # validate
    classifier.eval()
    total_num = 0
    total_correct = 0
    with torch.no_grad():
        for x, y, _ in dataloader_test:
            y_hat = classifier(x)
            y_hat = torch.nn.functional.softmax(y_hat, dim=1)
            # calculate number of correct predictions
            _, predicted = torch.max(y_hat, 1)
            num = predicted.shape[0]
            correct = (predicted == y).float().sum()
            total_num += num
            total_correct += correct
        acc = total_correct / total_num
        print('validation accuracy', acc)
    scheduler.step()
```
However, the classification performance on the downstream task is not as expected. Could you please advise on how I can adjust it? Thank you very much!

**Comment by guarin** [comment]
Hi, you might have to modify the classifier learning rate. Also make sure to inspect the pretraining and finetuning logs to check whether there are any anomalies and if the loss/accuracy are converging well. You might also want to plot the training accuracy and validation loss during finetuning for more insights.

**Comment by guarin** [comment]
Will close this for now, please reopen if you have more questions.
